pytorch_to_returnn/import_wrapper/wrap.py

```diff
@@ -51,6 +53,8 @@ def wrap(obj, *, name: str, ctx: WrapCtx):
       obj = make_wrapped_object(obj, name=name, ctx=ctx)
     elif not getattr(type(obj), "__module__", None) or not ctx.should_wrap_mod(type(obj).__module__):
       pass  # Don't wrap this object.
+    elif issubclass(type(obj), tuple):  # e.g. namedtuple
+      pass  # Don't wrap this object.
```

Comment:
I don't exactly understand this. Why is it a problem? This seems wrong.

Reply:
A `PackedSequence` should not be wrapped as a `WrappedObject`, and I excluded `namedtuples` here in general.

Reply:
But I don't understand this. Why should it not be wrapped? And why not `namedtuple`s in general?

Reply:
I think we discussed this to allow unpacking of `PackedSequence`s. I'm not sure about this, though. However, `make_wrapped_object` will fail for a `PackedSequence` in `_run_traced_orig_torch`:

```
  File "../pytorch-to-returnn/tests/test_layers.py", line 1302, in <module>
    line: globals()[arg]()  # assume function and execute
    locals:
      globals = <builtin> <built-in function globals>
      arg = <local> 'test_packed_sequence_1', len = 22
  File "../pytorch-to-returnn/tests/test_layers.py", line 362, in test_packed_sequence_1
    line: verify_torch_and_convert_to_returnn(model_func, inputs=x, inputs_data_kwargs={
              "shape": (None, n_feat), "time_dim_axis": 0, "batch_dim_axis": 1, "feature_dim_axis": 2})
    locals:
      verify_torch_and_convert_to_returnn = <global> <function verify_torch_and_convert_to_returnn at 0x7f933798a4c0>
      model_func = <local> <function test_packed_sequence_1.<locals>.model_func at 0x7f933798d9d0>
      inputs = <not found>
      x = <local> array([[[ 0.49671414, -0.1382643 ,  0.64768857,  1.5230298 ,
                           -0.23415338, -0.23413695,  1.5792128 ],
                          [ 0.7674347 , -0.46947438,  0.54256004, -0.46341768,
                           -0.46572974,  0.24196227, -1.9132802 ],
                          [-1.7249179 , -0.5622875 , -1.0128311 ,  0.31424734,
                           -0.9080..., len = 5, _[0]: {len = 3, _[0]: {len = 7}}
      inputs_data_kwargs = <not found>
      n_feat = <local> 7
  File "/u/vieting/testing/20210714_dfg_kickoff/pytorch-to-returnn/pytorch_to_returnn/converter/converter.py", line 328, in verify_torch_and_convert_to_returnn
    line: converter.run()
    locals:
      converter = <local> <pytorch_to_returnn.converter.converter.Converter object at 0x7f93cc825400>
      converter.run = <local> <bound method Converter.run of <pytorch_to_returnn.converter.converter.Converter object at 0x7f93cc825400>>
  File "/u/vieting/testing/20210714_dfg_kickoff/pytorch-to-returnn/pytorch_to_returnn/converter/converter.py", line 114, in Converter.run
    line: self._run_traced_orig_torch()
    locals:
      self = <local> <pytorch_to_returnn.converter.converter.Converter object at 0x7f93cc825400>
      self._run_traced_orig_torch = <local> <bound method Converter._run_traced_orig_torch of <pytorch_to_returnn.converter.converter.Converter object at 0x7f93cc825400>>
  File "/u/vieting/testing/20210714_dfg_kickoff/pytorch-to-returnn/pytorch_to_returnn/converter/converter.py", line 170, in Converter._run_traced_orig_torch
    line: out_wrapped = self._model_func(wrapped_import_torch_traced, wrapped_torch.from_numpy(self._inputs_np.copy()))
    locals:
      out_wrapped = <not found>
      self = <local> <pytorch_to_returnn.converter.converter.Converter object at 0x7f93cc825400>
      self._model_func = <local> <function test_packed_sequence_1.<locals>.model_func at 0x7f933798d9d0>
      wrapped_import_torch_traced = <global> <function wrapped_import_torch_traced at 0x7f933798a040>
      wrapped_torch = <local> <WrappedIndirectModule pytorch_to_returnn.import_wrapper._torch_traced.torch>
      wrapped_torch.from_numpy = <local> <class 'pytorch_to_returnn.import_wrapper.base_wrappers.function._VariableFunctionsClass.from_numpy'>
      self._inputs_np = <local> array([[[ 0.49671414, -0.1382643 ,  0.64768857,  1.5230298 ,
                                         -0.23415338, -0.23413695,  1.5792128 ],
                                        [ 0.7674347 , -0.46947438,  0.54256004, -0.46341768,
                                         -0.46572974,  0.24196227, -1.9132802 ],
                                        [-1.7249179 , -0.5622875 , -1.0128311 ,  0.31424734,
                                         -0.9080..., len = 5, _[0]: {len = 3, _[0]: {len = 7}}
      self._inputs_np.copy = <local> <built-in method copy of numpy.ndarray object at 0x7f93cc74da30>
  File "../pytorch-to-returnn/tests/test_layers.py", line 356, in test_packed_sequence_1.<locals>.model_func
    line: h = torch.nn.utils.rnn.pack_padded_sequence(inputs, [n_time] * n_batch)
    locals:
      h = <not found>
      torch = <local> <WrappedIndirectModule pytorch_to_returnn.import_wrapper._torch_traced.torch>
      torch.nn = <local> <WrappedIndirectModule pytorch_to_returnn.import_wrapper._torch_traced.torch.nn>
      torch.nn.utils = <local> <WrappedIndirectModule pytorch_to_returnn.import_wrapper._torch_traced.torch.nn.utils>
      torch.nn.utils.rnn = <local> <WrappedIndirectModule pytorch_to_returnn.import_wrapper._torch_traced.torch.nn.utils.rnn>
      torch.nn.utils.rnn.pack_padded_sequence = <local> <function pack_padded_sequence at 0x7f93379a00d0>
      inputs = <local> tensor([[[ 0.4967, -0.1383,  0.6477,  1.5230, -0.2342, -0.2341,  1.5792],
                                [ 0.7674, -0.4695,  0.5426, -0.4634, -0.4657,  0.2420, -1.9133],
                                [-1.7249, -0.5623, -1.0128,  0.3142, -0.9080, -1.4123,  1.4656]],

                               [[-0.2258,  0.0675, -1.4247, -0.5444,  0.1109, -1.1510,  0.3757],..., len = 5, _[0]: {len = 3, _[0]: {len = 7}}
      n_time = <local> 5
      n_batch = <local> 3
  File "/u/vieting/testing/20210714_dfg_kickoff/pytorch-to-returnn/pytorch_to_returnn/import_wrapper/base_wrappers/function.py", line 16, in _VariableFunctionsClass.from_numpy
    line: res = wrap(res, name="%s(...)" % name, ctx=ctx)
    locals:
      res = <local> PackedSequence(data=tensor([[ 0.4967, -0.1383,  0.6477,  1.5230, -0.2342, -0.2341,  1.5792],
                            [ 0.7674, -0.4695,  0.5426, -0.4634, -0.4657,  0.2420, -1.9133],
                            [-1.7249, -0.5623, -1.0128,  0.3142, -0.9080, -1.4123,  1.4656],
                            [-0.2258,  0.0675, -1.4247, -0.5444,  0.1109, -1.1..., len = 4, _[0]: {len = 15, _[0]: {len = 7}}
      wrap = <local> <function wrap at 0x7f933798dd30>
      name = <local> 'pytorch_to_returnn.import_wrapper._torch_traced.torch.nn.utils.rnn.pack_padded_sequence', len = 87
      ctx = <local> <WrapCtx 'pytorch_to_returnn.import_wrapper._torch_traced'>
  File "/u/vieting/testing/20210714_dfg_kickoff/pytorch-to-returnn/pytorch_to_returnn/import_wrapper/wrap.py", line 61, in wrap
    line: obj = make_wrapped_object(obj, name=name, ctx=ctx)
    locals:
      obj = <local> PackedSequence(data=tensor([[ 0.4967, -0.1383,  0.6477,  1.5230, -0.2342, -0.2341,  1.5792],
                            [ 0.7674, -0.4695,  0.5426, -0.4634, -0.4657,  0.2420, -1.9133],
                            [-1.7249, -0.5623, -1.0128,  0.3142, -0.9080, -1.4123,  1.4656],
                            [-0.2258,  0.0675, -1.4247, -0.5444,  0.1109, -1.1..., len = 4, _[0]: {len = 15, _[0]: {len = 7}}
      make_wrapped_object = <global> <function make_wrapped_object at 0x7f93379e7ca0>
      name = <local> 'pytorch_to_returnn.import_wrapper._torch_traced.torch.nn.utils.rnn.pack_padded_sequence(...)', len = 92
      ctx = <local> <WrapCtx 'pytorch_to_returnn.import_wrapper._torch_traced'>
  File "/u/vieting/testing/20210714_dfg_kickoff/pytorch-to-returnn/pytorch_to_returnn/import_wrapper/base_wrappers/object.py", line 88, in make_wrapped_object
    line: assert issubclass(cls, WrappedClassBase)
    locals:
      issubclass = <builtin> <built-in function issubclass>
      cls = <local> <class 'torch.nn.utils.rnn.PackedSequence'>
      WrappedClassBase = <local> <class 'pytorch_to_returnn.import_wrapper.base_wrappers.class_.WrappedClassBase'>
AssertionError
```

Reply:
You still have this here after our recent changes. Is this still needed?

Reply:
Yes, this is still needed. Without it, `make_wrapped_object` fails for a `PackedSequence` in `_run_traced_orig_torch` with the same error as above.

Reply:
But why? Why is it not wrapped? Or is it, and the type is messed up? Can you check `inspect.getmro(cls)`?

Or is this just together with the other change above, where you don't wrap types with `issubclass(obj, tuple)`? The error is gone when you remove both?

Reply:
Oh yes, it might be an interdependency of the two things. Now I get some other error, looking into this...

Reply:
If we have a `PackedSequence` as input to a module (here `LSTM`), there is a mismatch in the flattened inputs for `_run_traced_orig_torch()` and `_run_torch_returnn_drop_in()`. In the first case, the input is of type `pytorch_to_returnn.import_wrapper._torch_traced.torch.nn.utils.rnn.PackedSequence` which does not get flattened by `nest.flatten()`. In the second case, the input is of type `pytorch_to_returnn.torch.nn.modules.rnn.PackedSequence` which gets flattened into `[Tensor, None, None, None]`. Later, a check that both flattened inputs have the same lengths will fail.

Reply:
> `pytorch_to_returnn.import_wrapper._torch_traced.torch.nn.utils.rnn.PackedSequence` which does not get flattened by `nest.flatten()`

Why?

Reply:
I suppose because `pytorch_to_returnn.import_wrapper._torch_traced.torch.nn.utils.rnn.PackedSequence` is not a subclass of `tuple` and this is because `WrappedClass.__bases__ += (cls,)` in `make_wrapped_class()` throws an error which is ignored by the following exception (`TypeError: __bases__ assignment: 'PackedSequence' object layout differs from 'WrappedClassBase'`).